web/src/components/l10n/InstallerKeymapSwitcher.jsx

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) [2023] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import React from "react";
+import { FormSelect, FormSelectOption } from "@patternfly/react-core";
+
+import cockpit from "../../lib/cockpit";
+
+import { Icon } from "../layout";
+import { _ } from "~/i18n";
+import { useInstallerL10n } from "~/context/installerL10n";
+import { useL10n } from "~/context/l10n";
+import { localConnection } from "~/utils";
+
+const sort = (keymaps) => {
+  // sort the keymap names using the current locale
+  const lang = cockpit.language || "en";
+  return keymaps.sort((k1, k2) => k1.name.localeCompare(k2.name, lang));
+};
+
+export default function InstallerKeymapSwitcher() {
+  const { keymap, changeKeymap } = useInstallerL10n();
+  const { keymaps } = useL10n();
+
+  const onChange = (_, id) => changeKeymap(id);
+
+  const options = sort(keymaps)
+    .map((keymap, index) => <FormSelectOption key={index} value={keymap.id} label={keymap.name} />);
+
+  return (
+    <>
+      <h3>
+        {/* TRANSLATORS: label for keyboard layout selection */}
+        <Icon name="keyboard" size="24" />{_("Keyboard")}
+      </h3>
+      {(localConnection() &&
+      <FormSelect
+        id="keyboard"
+        // TRANSLATORS: label for keyboard layout selection
+        aria-label={_("keyboard")}
+        value={keymap}
+        onChange={onChange}
+      >
+        {options}
+      </FormSelect>) ||
+      // TRANSLATORS:
+      _("Keyboard layout cannot be changed in remote installation") }
```

Comment:
As suggested by @dgdavid during the review, I would show the current keymap as disabled for remote installation.

Reply:
> As suggested by @dgdavid during the review, I would show the current keymap as disabled for remote installation.

Just displaying it in plain text is enough along with the proper explanation in the popover @lslezak is working on. I mean, not needed to put a disabled selector there IMHO.

That said, at this moment I do not have an strong opinion for choosing between these two options.

Reply:
I tried that and I didn't like it. There are several problems:

- We would need to describe why the keyboard layout selector is disabled. It is not obvious and that could be frustrating for the users.
- It does not display your current local keyboard layout, which will be very likely different than the "English US" default.
- You do not care what's the keyboard layout on the remote machine, the current value is irrelevant, it does not provide any information for the remote user

See the screenshot:
![keyboard_remote9](https://github.com/openSUSE/agama/assets/907998/7909b6a6-e920-42a0-9e29-870b187d47ff)

Reply:
IMHO, one of the problem you found it's because using a disable selector for something that will be not be a selector. That's why I proposed using plain text and moving the explanation to a popover as we're doing now with the Language explanation.

But let's move on, we can re-evaluate at any time.

Reply:
As I understood it, the idea was to show the local keymap instead of the keymap in the remote system running Agama. But I guess that is not possible, right?

Reply:
Unfortunately that's not possible, you cannot get the local keyboard layout name in Javascript running in browser.

There is some experimental [Keyboard API](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard_API), but it does not provide the functionality we need.

You can get the name of a physical key to get a specified character, e.g. you can find which button produces letter `Y`. On the usual keyboard it is `Y`, but if you use QWERTZ layout you should get `Z`. (BTW I tested that with Chrome in KDE and it does not work properly with CZ QWERTZ layout.)

With that we could possibly only distinguish between QWERTY/QWERTZ/AZERTY variants, but we cannot get the full layout name like "Czech QWERTY".